### PR TITLE
support for observing metrics collected on seata client side

### LINKED ISSUE #4637: Metrics statistics of Seata Client SDK
Metrics statistics of Seata Client SDK

Description: 
- metrics statistics help users effectively monitor the execution process and identify possible problems.

Tasks: 
- The main monitoring indicators on the client SDK side include: Global begin / commit / rollback, branch registry / flush undo / report / commit / rollback and other status statistics.

- In addition, the calls to hooks such as undolog and fence should also be counted by metrics.

Programming language: 
- Java 

--- 

Seata client SDK侧的metrics统计

描述：
- metrics统计有助于帮助使用对Seata的执行过程进行有效的监控，帮助发现可能存在的问题。

任务：
-  客户端SDK侧的主要监控指标包括：Global begin/commit/rollback， Branch registry/flush undolog/report/commit/rollback等状态统计。
- 另外，对undolog，fence等hook的调用也要做metrics统计。

编程语言：
- Java

### PR DESCRIPTION
**Pull request by robynron**
- [x] I have registered the PR [changes](https://github.com/seata/seata/tree/develop/changes).

### Ⅰ. Describe what this PR did

Improved the monitoring of various data on the Seata Client side. Help users effectively monitor the execution process and identify possible problems.
### Ⅱ. Does this pull request fix one issue?

fixes #4637 

### Ⅲ. Why don't you add test cases (unit test/integration test)? 
 I'll add later.

### CODE REVIEW

File: metrics/seata-metrics-service/src/main/java/io/seata/metrics/TMMeterIdConstants.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
+package io.seata.metrics;
```
Comment (lvekee) [inline metrics/seata-metrics-service/src/main/java/io/seata/metrics/TMMeterIdConstants.java:1]: Copyright 加一下

File: metrics/seata-metrics-service/src/main/java/io/seata/metrics/service/MetricsPublisher.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
+package io.seata.metrics.service;
```
Comment (lvekee) [inline metrics/seata-metrics-service/src/main/java/io/seata/metrics/service/MetricsPublisher.java:1]: Copyright加一下

File: metrics/seata-metrics-service/src/main/java/io/seata/metrics/service/MetricsPublisher.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+import io.seata.common.ConfigurationKeys;
+import io.seata.config.ConfigurationFactory;
+import io.seata.core.context.RootContext;
```
Comment (lvekee) [inline metrics/seata-metrics-service/src/main/java/io/seata/metrics/service/MetricsPublisher.java:5]: 无用包不要导入

File: metrics/seata-metrics-service/src/main/java/io/seata/metrics/service/MetricsSubscriber.java [hunk 4]
```diff
@@ -0,0 +1,771 @@
...
+import io.seata.metrics.registry.Registry;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+
+import static io.seata.metrics.IdConstants.*;
```
Comment (lvekee) [inline metrics/seata-metrics-service/src/main/java/io/seata/metrics/service/MetricsSubscriber.java:36]: 不要*号导入

File: metrics/seata-metrics-service/src/main/java/io/seata/metrics/service/MetricsSubscriber.java [hunk 5]
```diff
@@ -0,0 +1,771 @@
...
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package io.seata.metrics.service;
+
+import com.google.common.eventbus.Subscribe;
+import io.seata.core.event.BranchEvent;
+import io.seata.core.event.GlobalTransactionEvent;
+import io.seata.core.model.GlobalStatus;
+import io.seata.metrics.IdConstants;
```
Comment (lvekee) [inline metrics/seata-metrics-service/src/main/java/io/seata/metrics/service/MetricsSubscriber.java:22]: 无用包不要导入

File: metrics/seata-metrics-service/src/main/java/io/seata/metrics/service/MetricsSubscriber.java [hunk 6]
```diff
@@ -0,0 +1,771 @@
...
+            consumers.put(GlobalStatus.RollbackFailed.name(), this::processGlobalStatusRollbackFailed);
+            consumers.put(GlobalStatus.TimeoutRollbacked.name(), this::processGlobalStatusTimeoutRollbacked);
+            consumers.put(GlobalStatus.TimeoutRollbackFailed.name(), this::processGlobalStatusTimeoutRollbackFailed);
+
+            consumers.put(GlobalStatus.CommitRetryTimeout.name(), this::processGlobalStatusCommitRetryTimeout);
+            consumers.put(GlobalStatus.RollbackRetryTimeout.name(), this::processGlobalStatusTimeoutRollbackRetryTimeout);
+
+            consumers.put(STATUS_VALUE_AFTER_COMMITTED_KEY, this::processAfterGlobalCommitted);
+            consumers.put(STATUS_VALUE_AFTER_ROLLBACKED_KEY, this::processAfterGlobalRollbacked);
+        }else if (MetricsManager.get().getRole().equals(MetricsManager.get().ROLE_VALUE_CLIENT)) {
```
Comment (lvekee) [inline metrics/seata-metrics-service/src/main/java/io/seata/metrics/service/MetricsSubscriber.java:69]: p3c格式化下代码

File: metrics/seata-metrics-service/src/main/java/io/seata/metrics/service/MetricsSubscriber.java [hunk 7]
```diff
@@ -0,0 +1,771 @@
...
+        registry.getSummary(TCMeterIdConstants.SUMMARY_TWO_PHASE_TIMEOUT
+                .withTag(APP_ID_KEY, event.getApplicationId())
+                .withTag(GROUP_KEY, event.getGroup())).increase(1);
+    }
+
+
+    @Subscribe
+    public void recordGlobalTransactionEventForMetrics(GlobalTransactionEvent event) {
+        if (registry != null) {
+            if (consumers.containsKey(event.getMetricEvent())){
```
Comment (lvekee) [inline metrics/seata-metrics-service/src/main/java/io/seata/metrics/service/MetricsSubscriber.java:742]: 格式化下代码 少空格

File: metrics/seata-metrics-service/src/main/java/io/seata/metrics/RMMeterIdConstants.java [hunk 8]
```diff
@@ -0,0 +1,327 @@
+package io.seata.metrics;
+
```
Comment (lvekee) [inline metrics/seata-metrics-service/src/main/java/io/seata/metrics/RMMeterIdConstants.java:2]: CopyRight加一下

File: metrics/seata-metrics-service/src/main/java/io/seata/metrics/TCMeterIdConstants.java [hunk 9]
```diff
@@ -13,17 +13,17 @@
  *  See the License for the specific language governing permissions and
  *  limitations under the License.
  */
-package io.seata.server.metrics;
+package io.seata.metrics;
 
-import io.seata.metrics.IdConstants;
 import io.seata.metrics.Id;
+import io.seata.metrics.IdConstants;
```
Comment (lvekee) [inline metrics/seata-metrics-service/src/main/java/io/seata/metrics/TCMeterIdConstants.java:19]: 导包改一下

**Review by lvekee — CHANGES_REQUESTED** [review]

File: rm/src/main/java/io/seata/rm/DefaultResourceManager.java [hunk 10]
```diff
@@ -28,6 +28,7 @@
 import io.seata.core.model.BranchType;
 import io.seata.core.model.Resource;
 import io.seata.core.model.ResourceManager;
+import io.seata.metrics.service.MetricsPublisher;
```
Comment (lvekee) [inline rm/src/main/java/io/seata/rm/DefaultResourceManager.java:31]: 去掉无用导包

File: rm/src/main/java/io/seata/rm/DefaultResourceManager.java [hunk 11]
```diff
@@ -112,6 +113,7 @@ public boolean lockQuery(BranchType branchType, String resourceId,
     @Override
     public void registerResource(Resource resource) {
         getResourceManager(resource.getBranchType()).registerResource(resource);
+
```
Comment (lvekee) [inline rm/src/main/java/io/seata/rm/DefaultResourceManager.java:116]: 去掉空行

**Review by lvekee — CHANGES_REQUESTED** [review]

**Comment by Bughue** [comment]
是不是要关联下  https://github.com/seata/seata/tree/develop/metrics#readme

**Comment by funky-eyes** [comment]
Due to code conflicts and age, if you are still interested in participating in the community, please resubmit this pr to the 2.x branch after resolving the conflict.
由于代码冲突以及年代久远，如果你还有兴趣参与到社区，请解决完冲突后重新请提交这份pr至2.x分支中

**Comment by robynron** [comment]
好的，我后面基于2.x分支重新弄一下

**Pull request closed without merging**